src/audio/pcm_converter/pcm_converter_generic.c

```diff
@@ -817,6 +817,8 @@ const struct pcm_func_vc_map pcm_func_vc_map[] = {
 		ipc4_gtw_host, ipc4_playback, pcm_convert_s32_to_s24 },
 	{ SOF_IPC_FRAME_S32_LE, SOF_IPC_FRAME_S24_4LE, SOF_IPC_FRAME_S32_LE, SOF_IPC_FRAME_S24_4LE,
 		ipc4_gtw_host, ipc4_capture, pcm_convert_s24_to_s32 },
+	{ SOF_IPC_FRAME_S32_LE, SOF_IPC_FRAME_S24_4LE, SOF_IPC_FRAME_S32_LE, SOF_IPC_FRAME_S32_LE,
+		ipc4_gtw_all, ipc4_bidirection, pcm_convert_s24_to_s32},
```

Comment:
Will this work correctly for case when gateway is host and direction is playback? I think in that case audio comes as 24 bit in upper 3 bytes with 0 in lower byte. Shifting that data left will result in samples corruption.

Reply:
@serhiy-katsyuba-intel  I validated sound output on Linux platform. On Linux the S24_4LE sample bits is in low 3 bytes, but on windows I got data in high 24bits.  I mentioned that windows sends different data layout compared to your test case, now I think it should be different.

I checked FW code again and confirm I can't find anything wrong in dai setting.  According to HDA spec, the 24bit sample data on HDA bus should be in high 24 bit. I checked all the commits in history and confirm that our gateway doesn't compatible with it so we need to do some conversion in dai component, but not in host. So we don't need to do some conversion in host dma for this case. And you can submit your patch again which should be correct to me.

I will check windows data layout again.

Reply:
I think for Windows we can call the 24bit (MSB) format as 32bit (since the LSBs are ignored). i.e. the code should internally select 32bit when Windows asks for the 24bit MSB format.